### DISCUSSION TITLE
SmartAutoClicker-2.3.0

### DISCUSSION THREAD

**Discussion by Nain57** [Nain57/Smart-AutoClicker#312] [Announcements]
## Detection algorithm rework
This version contains a complete rework of the detection algorithm. I have been more rigorous with my test protocol lately so I have been able to measure the improvements.

The protocol is the following:

- 3 images (from my OnePlus 10 Pro device,  3216 x 1440): one small (the size of a launcher icon), one medium (the size of 4 launcher icons in width, 1 in height), one big (the width and height of a 4x4 grid of launcher icons
- Each of this image is used in 2 conditions, on detecting at position, one on the whole screen
- Detection quality to maximum (1200)
- Detect during 10000 screen frames

Here are the results:
* RAM usage went from 332Mb max to 161Mb max
* Success went from 8451 to 10000 (100% of correct detections)
* Average detection time is divided by 10 for EXACT detection, and by 5 for WHOLE SCREEN.

As you can see, the results are really great, and the improvements should be seen by the user. As a consequence, i have updated the Detection Quality feature, by increasing its minimum to 800, and the maximum to 3216 (which is, according to my tests, still faster than the old maximum). All your scenario detection qualities will be automatically updated and increased in order to improve automatically all your scenario.

## Condition operator and click on condition
This feature wasn't really clear, you never know on which condition it will click. This is now fixed by adding a condition selector to the click action in the case the event condition operator is set to All conditions.

## Tutorial update
A new tutorial has been added, and all of them have been translated in the supported languages.

## Stabilization
A lot of corner cases have been fixed in this version, and the detection algorithm revamp should also be more robust. 
The user crash free rate on the 2.1.4 version is 99.76%, let's try to reach 99.9% !

<hr /><em>This discussion was created from the release <a href='https://github.com/Nain57/Smart-AutoClicker/releases/tag/2.3.0'>SmartAutoClicker-2.3.0</a>.</em>

**Comment by stansu** [comment]
glad to see you back, thank you.

**Comment by xihh12138** [comment]
The detection efficiency and speed have indeed improved a lot ヾ(≧▽≦*)o，and the probability of matching has also increased, so the threshold of detection conditions should be reduced a bit to reduce the probability of mismatching []~(￣▽￣)~*

**Comment by DaMaddCyantist** [comment]
Awesome @Nain57 
Im gonna make the upgrade to this version and keep it as a new base. Cheers.

**Comment by c-sanchez** [comment]
The option to select whether I want to search for the condition in the exact place or in the whole screen + the option to click on a condition are really perfect.
I've recreated the macro/scenario I had using these functions now is a lot better.

You are amazing @Nain57
Thank you so very much.

**Comment by c-sanchez** [comment]
Btw, something that could be improved is that the condition/image, besides searching either in the exact place or in the whole screen, could be something in between, allowing to select a specific region to search.

Many times things move just a little bit of place, so even if it is not in the exact place, it would be in a very close area, so selecting a search area I guess would be much faster than searching the whole screen.

**Reply by DaMaddCyantist** [reply]
You can do this as of v2.3.2

**Reply by c-sanchez** [reply]
Yeah, I noticed that @Nain57 added it, it's great :D

**Reply by Nain57** [reply]
That was in my mind for a long time but it wasn't possible before the 2.3.0 detection algorithm revamp. With this new structure, this was a quick win for the project (it took me more time to add the UI than the actual feature). 
This new structure should allow new detections types, such as Text Detection (I have a working prototype, but it is still ages from a stable release, do not expect it before 2024)
